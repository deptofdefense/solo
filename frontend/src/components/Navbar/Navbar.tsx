import React, { useState, useCallback } from "react";
import {
  Header,
  HeaderNavbar,
  HeaderNav,
  HeaderNavLink,
  HeaderLogo
} from "solo-uswds";
import { useAuthContext } from "context";

const Navbar: React.FC = () => {
  const { authenticated } = useAuthContext();
  const [isOpen, setIsOpen] = useState(false);

  const toggleOpen = useCallback(() => {
    setIsOpen(!isOpen);
  }, [isOpen, setIsOpen]);

  const close = () => setIsOpen(false);

  return (
    <Header>
      <HeaderNavbar>
        <HeaderLogo text="SOLO" />
        <button onClick={toggleOpen} className="usa-menu-btn">
          Menu
        </button>
      </HeaderNavbar>
      <HeaderNav isOpen={isOpen} onClose={toggleOpen}>
        {authenticated && (
          <>
            <HeaderNavLink to="/" exact onClick={close}>
              Status
            </HeaderNavLink>
            <HeaderNavLink to="/d6t" exact onClick={close}>
              Enter Receipt (D6T)
            </HeaderNavLink>
            <HeaderNavLink to="/cor" exact onClick={close}>
              Confirmation of Receipt (COR)
            </HeaderNavLink>
          </>
        )}
      </HeaderNav>
    </Header>
  );
};

export default Navbar;
