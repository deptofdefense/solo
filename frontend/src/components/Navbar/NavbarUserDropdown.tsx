import React, { useState } from "react";
import classNames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faUser, faUserLock } from "@fortawesome/free-solid-svg-icons";
import classes from "./NavbarUserDropdown.module.scss";

interface NavbarUserDropdownProps {
  username: string;
  onLogout: () => void;
}

const NavbarUserDropdown: React.FC<NavbarUserDropdownProps> = ({
  username,
  onLogout
}) => {
  const [expanded, setExpanded] = useState(false);

  const onLogoutClick: React.MouseEventHandler<HTMLAnchorElement> = event => {
    event.preventDefault();
    onLogout();
  };

  return (
    <li className="usa-nav__primary-item">
      <button
        onClick={() => setExpanded(!expanded)}
        className={classNames("usa-accordion__button", "usa-nav__link", {
          "usa-current": expanded
        })}
        aria-expanded={expanded}
      >
        <FontAwesomeIcon className="margin-right-1" icon={faUser} />
        <span>{username}</span>
      </button>
      {expanded && (
        <ul className={classNames("usa-nav__submenu", classes.submenu)}>
          <li className="usa-nav__submenu-item">
            <a href="/postlogout" onClick={onLogoutClick}>
              <FontAwesomeIcon className="margin-right-1" icon={faUserLock} />
              Logout
            </a>
          </li>
        </ul>
      )}
    </li>
  );
};

export default NavbarUserDropdown;
