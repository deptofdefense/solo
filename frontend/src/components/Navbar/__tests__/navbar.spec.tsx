import React from "react";
import { render, fireEvent, wait } from "test-utils";
import Navbar from "../Navbar";

describe("Navbar component", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("clicking menu adds visibility class", async () => {
    const { container, getByText } = render(<Navbar />);
    const primaryNavSelector = '[aria-label="Primary navigation"]';
    expect(container.querySelector(primaryNavSelector)).not.toHaveClass(
      "is-visible"
    );
    const menuButton = getByText("Menu");
    fireEvent.click(menuButton);
    await wait(() => {
      expect(container.querySelector(primaryNavSelector)).toHaveClass(
        "is-visible"
      );
    });
  });

  it("clicking a link closes the menu", async () => {
    const { container, getByText } = render(<Navbar />);
    const primaryNavSelector = '[aria-label="Primary navigation"]';
    const menuButton = getByText("Menu");
    fireEvent.click(menuButton);
    await wait(() => {
      expect(container.querySelector(primaryNavSelector)).toHaveClass(
        "is-visible"
      );
    });
    const aboutLink = getByText("About");
    fireEvent.click(aboutLink);
    await wait(() => {
      expect(container.querySelector(primaryNavSelector)).not.toHaveClass(
        "is-visible"
      );
    });
  });
});
