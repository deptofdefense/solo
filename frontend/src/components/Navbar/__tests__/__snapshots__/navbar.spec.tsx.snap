// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="usa-header usa-header--basic"
  >
    <div
      class="usa-nav-container navContainer"
    >
      <div
        class="usa-navbar"
      >
        <div
          class="usa-logo logoHeader"
        >
          <em
            class="usa-logo__text"
          >
            <a
              aria-current="page"
              class="active"
              href="/"
            >
              <img
                alt="SOLO"
                class="logoImg"
                src="solo.png"
              />
            </a>
          </em>
        </div>
        <button
          class="usa-menu-btn"
        >
          Menu
        </button>
      </div>
      <nav
        aria-label="Primary navigation"
        class="usa-nav"
      >
        <ul
          class="usa-nav__primary usa-accordion"
        >
          <button
            class="usa-nav__close"
          >
            <img
              alt="close"
              src="close.svg"
            />
          </button>
        </ul>
      </nav>
    </div>
  </header>
</DocumentFragment>
`;
